modules/tensor_mechanics/src/materials/RecomputeRadialReturnStressIncrement.C

```diff
@@ -0,0 +1,121 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#include "RecomputeRadialReturnStressIncrement.h"
+
+template<>
+InputParameters validParams<RecomputeRadialReturnStressIncrement>()
+{
+  InputParameters params = validParams<RecomputeGeneralReturnStressIncrement>();
+
+  // Newton Iteration control parameters
+  params.addParam<bool>("output_iteration_info", false, "Set true to output newton iteration information from the radial return material");
+  params.addParam<bool>("output_iteration_info_on_error", false, "Set true to output the recompute material iteration information when a step fails");
+  params.addParam<Real>("relative_tolerance", 1e-8, "Relative convergence tolerance for the newton iteration within the radial return material");
+  params.addParam<Real>("absolute_tolerance", 1e-20, "Absolute convergence tolerance for newton iteration within the radial return material");
+  return params;
+}
+
+RecomputeRadialReturnStressIncrement::RecomputeRadialReturnStressIncrement(const InputParameters & parameters) :
+    RecomputeGeneralReturnStressIncrement(parameters),
+    _output_iteration_info(getParam<bool>("output_iteration_info")),
+    _output_iteration_info_on_error(getParam<bool>("output_iteration_info_on_error")),
+    _relative_tolerance(parameters.get<Real>("relative_tolerance")),
+    _absolute_tolerance(parameters.get<Real>("absolute_tolerance"))
+{
+}
+
+void
+RecomputeRadialReturnStressIncrement::computeStressIncrement()
+{
+  // Given the stretching, compute the stress increment and add iteration to the old stress. Also update the inelastic strain
+  // Compute the stress in the intermediate configuration while retaining the stress history
+  _return_stress_increment[_qp] = _elasticity_tensor[_qp] * _strain_increment[_qp] + _stress_old[_qp];
+
+  // compute the deviatoric trial stress and trial strain from the current intermediate configuration
+  RankTwoTensor deviatoric_trial_stress = _return_stress_increment[_qp].deviatoric();
+
+  // compute the effective trial stress
+  Real dev_trial_stress_squared = deviatoric_trial_stress.doubleContraction(deviatoric_trial_stress);
+  Real effective_trial_stress = std::sqrt(3.0 / 2.0 * dev_trial_stress_squared);
+
+  //If the effective trial stress is zero, so should the inelastic strain increment be zero
+  //In that case skip the entire iteration if the effective trial stress is zero
+  if (effective_trial_stress != 0.0)
+  {
+    computeStressInitialize(effective_trial_stress);
```

Comment:
Rather than setting the inelastic strain equal to zero at the end of the radial return iteration, I decided to check for a non-zero effective trial stress before running the while loop when applying Daniel's recommendation.  Are there use cases where we would want to go through the while loop iteration with a zero effective stress?